scrapyd_k8s/joblogs/log_handler_k8s.py

```diff
@@ -0,0 +1,281 @@
+import os
+import threading
+import tempfile
+import logging
+
+from kubernetes import client, watch
+from scrapyd_k8s.object_storage import LibcloudObjectStorage
+
+logger = logging.getLogger(__name__)
+
+class KubernetesJobLogHandler:
+    """
+    A class to handle Kubernetes job logs by watching pods, streaming logs, and uploading them to object storage.
+
+    ...
+
+    Attributes
+    ----------
+    DEFAULT_BLOCK_SIZE : int
+        Default size (in bytes) of blocks to read when retrieving the last N lines from a file.
+    config : object
+        Configuration object containing settings for job logs and storage.
+    watcher_threads : dict
+        Dictionary to keep track of watcher threads for each pod.
+    pod_tmp_mapping : dict
+        Mapping of pod names to their temporary log file paths.
+    namespace : str
+        Kubernetes namespace to watch pods in.
+    num_lines_to_check : int
+        Number of lines to check for matching logs when resuming log streaming.
+    object_storage_provider : LibcloudObjectStorage
+        Instance of the object storage provider for uploading logs.
+
+    Methods
+    -------
+    start():
+        Starts the pod watcher thread for job logs.
+    get_last_n_lines(file_path, num_lines):
+        Efficiently retrieves the last `num_lines` lines from a file.
+    concatenate_and_delete_files(main_file_path, temp_file_path, block_size=6144):
+        Concatenates a temporary file to the main log file and deletes the temporary file.
+    make_log_filename_for_job(job_name):
+        Generates a unique temporary file path for storing logs of a job.
+    stream_logs(job_name):
+        Streams logs from a Kubernetes pod and writes them to a file.
+    watch_pods():
+        Watches Kubernetes pods and handles events such as starting log streaming or uploading logs.
+    """
+    # The value was chosen to provide a balance between memory usage and the number of I/O operations
+    DEFAULT_BLOCK_SIZE = 6144
+
+    def __init__(self, config):
+        """
+        Constructs all the necessary attributes for the KubernetesJobLogHandler object.
+
+        Parameters
+        ----------
+        config : object
+            Configuration object containing settings for job logs and storage.
+        """
+        self.config = config
+        self.watcher_threads = {}
+        self.pod_tmp_mapping = {}
+        self.namespace = config.namespace()
+        self.num_lines_to_check = int(config.joblogs().get('num_lines_to_check', 0))
+        self.object_storage_provider = None
```

Comment:
> Also, what happens if num_lines_to_check cannot be cast to an integer?
In case a user completely ignored documentation and comments in config, the code will raise TypeError and stop working. This is a desireable behavior.

> Some of the attributes can also be defined as class attributes
If a user wants to initialize two instances with different object providers then the code won't work properly, so I prefer to keep essential, variable and mutable parameters as instance attributes, not class since each instance has its own configuration and may vary.